@xen-orchestra/lite/src/components/Console.tsx

```diff
@@ -38,47 +43,99 @@ interface Computed {}
 const Console = withState<State, Props, Effects, Computed, ParentState, ParentEffects>(
   {
     initialState: () => ({
+      _timeout: undefined,
       container: React.createRef(),
       rfb: undefined,
+      rfbConnected: false,
     }),
     effects: {
-      initialize: async function () {
+      initialize: function () {
+        this.effects._connect()
+      },
+      _handleDisconnect: function () {
+        this.state.rfbConnected = false
+        this.effects._connect()
+      },
+      _connect: async function () {
         const { vmId } = this.props
-        const { objectsByType, xapi } = this.state
-        const consoles = (objectsByType.get('VM')?.get(vmId) as Vm)?.$consoles.filter(
-          vmConsole => vmConsole.protocol === 'rfb'
-        )
-
-        if (consoles === undefined || consoles.length === 0) {
-          throw new Error('Could not find VM console')
+        const { objectsByType, rfb, xapi } = this.state
+
+        // 8 sequences mean 54S
+        for (const delay of fibonacci().toMs().take(8)) {
+          try {
+            const consoles = (objectsByType.get('VM')?.get(vmId) as Vm)?.$consoles.filter(
+              vmConsole => vmConsole.protocol === 'rfb'
+            )
+
+            if (rfb !== undefined) {
+              rfb.removeEventListener('connect', this.effects._displayConsole)
+              rfb.removeEventListener('disconnect', this.effects._handleDisconnect)
+            }
+
+            if (consoles === undefined || consoles.length === 0) {
+              throw new Error('Could not find VM console')
+            }
+
+            if (xapi.sessionId === undefined) {
+              throw new Error('Not connected to XAPI')
+            }
+
+            const url = new URL(consoles[0].location)
+            url.protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:'
+            url.searchParams.set('session_id', xapi.sessionId)
+
+            this.state.rfb = new RFB(this.state.container.current, url, {
+              wsProtocols: ['binary'],
+            })
+            this.state.rfb.addEventListener('connect', this.effects._displayConsole)
+            this.state.rfb.addEventListener('disconnect', this.effects._handleDisconnect)
+            this.state.rfb.scaleViewport = true
+            this.props.setCtrlAltDel(this.effects.sendCtrlAltDel)
+            return
+          } catch (error) {
+            await new Promise(resolve => (this.state._timeout = setTimeout(() => resolve(''), delay)))
+          }
         }
-
-        if (xapi.sessionId === undefined) {
-          throw new Error('Not connected to XAPI')
+        throw new Error('Unable to connect to the VM console. Too many attempts')
+      },
+      _displayConsole: function () {
+        this.state.rfbConnected = true
+      },
```

Comment:
Call it `_handleConnect`. If we ever want to add something to this function, the name won't make sense anymore.